doc/developer-guide/io-framework.md

```diff
@@ -0,0 +1,294 @@
+focused on the io_uring-based implementation. For differences when
+io_uring is not present, check section [Fallback mode](#fallback-mode)
+
+### Initialization
+
+The framework is initialized using `gf_io_run()`.
+
+```c
+typedef int32_t (*gf_io_async_t)(gf_io_op_t *op);
+
+typedef struct {
+    gf_io_async_t setup;
+    gf_io_async_t cleanup;
+} gf_io_handlers_t;
+
+int32_t gf_io_run(gf_io_handlers_t handlers, void *data);
+```
+
+The handlers structure contains two async functions, one that is called
+just after having initialized the I/O infrastructure, and another one that
+is called after stopping everything else. The 'data' argument is an extra
+argument that will be passed to each function.
+
+The returned value can be a negative error code if there has been any
+problem while initializing the system, or 0 if everything worked fine. In
+this case, the function only returns when the program is terminating.
+
+### Termination
+
+When it's determined that the process must be terminated, a call to
+`gf_io_shutdown()` must be done.
+
+```c
+void gf_io_shutdown(void);
+```
+
+This function initiates a shutdown procedure, but returns immediately.
+Once the shutdown is completed, `gf_io_run()` will return. It can be
+called from anywhere.
+
+When shutdown is initiated, all I/O should have been stopped. If there
+is active I/O during the shutdown, they can complete, fail or be cancelled,
+depending on what state the request was. To ensure consistent behavior, try
+to always stop I/O before terminating the I/O framework.
+
+### Normal operation
+
+After everything is ready, the normal operation of the I/O framework is
+very simple:
+
+1. A worker picks one completion event from the kernel.
+
+2. The callback associated to the completion event is executed.
+
+   2.1. The callback can prepare new I/O requests using one of the
+        `gf_io_*` I/O functions available for I/O operations.
+
+   2.2. Requests can be sent one by one or submitted in a batch. In all
+        cases they are added to the io_uring SQ ring.
+
+   2.3. The callback may (but it's not required to) explicitly flush all
+        queued requests to the kernel using `gf_io_flush()`.
+
+3. Once the callback finishes, any queued requests (from this worker or
+   any other worker that has added requests to the queue) are automatically
+   flushed.
+
+### Available I/O operations
+
+The I/O framework supports two ways of sending operations to the kernel.
+In direct mode, each request is sent independently of the others. In batch
+mode multiple requests are sent together all at once.
+
+All operations will also have a `data` argument to pass any additional
+per-request private data that the callback may need. This data will be
+available in `op->data` for most of the cases (there's an exception for
+asynchronous requests. See later).
+
+Many of the I/O operations will have a timeout argument, which represents
+the maximum time allowed for the I/O to complete. If the operation takes
```

Comment:
The timeout can be used to automatically cancel pending operations (frame timeout would be one example, but it could also be used to cancel writes taking too much time or other I/O operations that should finish fast (probably this will also need some kernel support that I think is not present right now)), but they will also be the core functionality to implement timers in Gluster.

Conceptually it could even be used to cancel callbacks taking too long, but this requires an infrastructure that is not present and it's not easy to implement correctly. For now timeouts will be ignored when not supported (for example the legacy mode of the I/O framework already ignores any timeout).